File: bindings/cpp/tests/test_sig.cpp
/*
 * libpqc-dyber - Post-Quantum Cryptography Library
 * Copyright (c) 2024-2026 Dyber, Inc.
 * SPDX-License-Identifier: Apache-2.0 OR MIT
 *
 * Basic tests for the C++ Signature wrapper.
 */

#include <cassert>
#include <cstdio>
#include <string>

#include <pqc/pqc.hpp>

static const pqc::Bytes test_message = {
    'T', 'h', 'i', 's', ' ', 'i', 's', ' ', 'a', ' ',
    't', 'e', 's', 't', ' ', 'm', 'e', 's', 's', 'a',
    'g', 'e', '.',
};

static void test_sig_properties() {
    pqc::Signature sig("ML-DSA-65");

    assert(sig.algorithm() == "ML-DSA-65");
    assert(sig.public_key_size() > 0);
    assert(sig.secret_key_size() > 0);
    assert(sig.max_signature_size() > 0);
    assert(!sig.is_stateful());

    std::printf("[PASS] test_sig_properties\n");
}

static void test_sig_sign_verify() {
    pqc::Signature sig("ML-DSA-65");

    auto [pk, sk] = sig.keygen();
    assert(pk.size() == sig.public_key_size());
    assert(sk.size() == sig.secret_key_size());

    auto signature = sig.sign(test_message, sk);
    assert(!signature.empty());
    assert(signature.size() <= sig.max_signature_size());

    bool valid = sig.verify(test_message, signature, pk);
    assert(valid);

    std::printf("[PASS] test_sig_sign_verify\n");
}

static void test_sig_verify_tampered() {
    pqc::Signature sig("ML-DSA-65");

    auto [pk, sk] = sig.keygen();
    auto signature = sig.sign(test_message, sk);

    // Tamper with the message
    pqc::Bytes tampered = test_message;
    tampered[0] ^= 0xFF;

    bool valid = sig.verify(tampered, signature, pk);
    assert(!valid);

    std::printf("[PASS] test_sig_verify_tampered\n");
}

static void test_sig_invalid_algorithm() {
    bool caught = false;
    try {
        pqc::Signature sig("INVALID-ALGORITHM-NAME");
    } catch (const pqc::Error& e) {
        caught = true;
        assert(e.status() == PQC_ERROR_NOT_SUPPORTED);
    }
    assert(caught);

    std::printf("[PASS] test_sig_invalid_algorithm\n");
}

static void test_sig_move() {
    pqc::Signature sig1("ML-DSA-65");
    auto [pk, sk] = sig1.keygen();

    pqc::Signature sig2(std::move(sig1));
    assert(sig2.algorithm() == "ML-DSA-65");

    auto signature = sig2.sign(test_message, sk);
    bool valid = sig2.verify(test_message, signature, pk);
    assert(valid);

    std::printf("[PASS] test_sig_move\n");
}

static void test_sig_algorithm_listing() {
    auto algorithms = pqc::sig_algorithms();
    assert(!algorithms.empty());

    bool found_mldsa = false;
    for (const auto& name : algorithms) {
        if (name == "ML-DSA-65") {
            found_mldsa = true;
        }
    }
    assert(found_mldsa);
    assert(pqc::sig_is_enabled("ML-DSA-65"));

    std::printf("[PASS] test_sig_algorithm_listing\n");
}

static void test_sig_raw_pointer_api() {
    pqc::Signature sig("ML-DSA-65");
    auto [pk, sk] = sig.keygen();

    const uint8_t raw_msg[] = "raw pointer message";
    std::size_t raw_len = sizeof(raw_msg) - 1;

    auto signature = sig.sign(raw_msg, raw_len, sk);
    bool valid = sig.verify(raw_msg, raw_len, signature, pk);
    assert(valid);

    std::printf("[PASS] test_sig_raw_pointer_api\n");
}

int main() {
    pqc::LibraryGuard guard;

    test_sig_properties();
    test_sig_sign_verify();
    test_sig_verify_tampered();
    test_sig_invalid_algorithm();
    test_sig_move();
    test_sig_algorithm_listing();
    test_sig_raw_pointer_api();

    std::printf("\nAll Signature tests passed.\n");
    return 0;
}

File: bindings/python/tests/test_sig.py
# libpqc-dyber - Post-Quantum Cryptography Library
# Copyright (c) 2024-2026 Dyber, Inc.
# SPDX-License-Identifier: Apache-2.0 OR MIT

"""Pytest tests for the Signature binding."""

import pytest

from pqc_dyber import (
    AlgorithmNotSupportedError,
    PQCError,
    Signature,
    sig_algorithms,
    sig_is_enabled,
)


TEST_MESSAGE = b"This is a test message."


class TestSignatureProperties:
    """Test Signature context properties."""

    def test_algorithm_name(self):
        sig = Signature("ML-DSA-65")
        assert sig.algorithm == "ML-DSA-65"

    def test_sizes_are_positive(self):
        sig = Signature("ML-DSA-65")
        assert sig.public_key_size > 0
        assert sig.secret_key_size > 0
        assert sig.max_signature_size > 0

    def test_is_not_stateful(self):
        sig = Signature("ML-DSA-65")
        assert sig.is_stateful is False

    def test_repr(self):
        sig = Signature("ML-DSA-65")
        assert repr(sig) == "Signature('ML-DSA-65')"

    def test_invalid_algorithm_raises(self):
        with pytest.raises(AlgorithmNotSupportedError):
            Signature("INVALID-ALGORITHM-NAME")


class TestSignatureOperations:
    """Test sign / verify round-trip."""

    def test_sign_verify_mldsa65(self):
        sig = Signature("ML-DSA-65")
        pk, sk = sig.keygen()

        assert isinstance(pk, bytes)
        assert isinstance(sk, bytes)
        assert len(pk) == sig.public_key_size
        assert len(sk) == sig.secret_key_size

        signature = sig.sign(TEST_MESSAGE, sk)
        assert isinstance(signature, bytes)
        assert len(signature) <= sig.max_signature_size
        assert len(signature) > 0

        assert sig.verify(TEST_MESSAGE, signature, pk) is True

    def test_sign_verify_mldsa44(self):
        sig = Signature("ML-DSA-44")
        pk, sk = sig.keygen()
        signature = sig.sign(TEST_MESSAGE, sk)
        assert sig.verify(TEST_MESSAGE, signature, pk) is True

    def test_sign_verify_mldsa87(self):
        sig = Signature("ML-DSA-87")
        pk, sk = sig.keygen()
        signature = sig.sign(TEST_MESSAGE, sk)
        assert sig.verify(TEST_MESSAGE, signature, pk) is True

    def test_verify_tampered_message(self):
        sig = Signature("ML-DSA-65")
        pk, sk = sig.keygen()
        signature = sig.sign(TEST_MESSAGE, sk)

        tampered = b"Tampered message."
        assert sig.verify(tampered, signature, pk) is False

    def test_verify_tampered_signature(self):
        sig = Signature("ML-DSA-65")
        pk, sk = sig.keygen()
        signature = sig.sign(TEST_MESSAGE, sk)

        tampered_sig = bytearray(signature)
        tampered_sig[0] ^= 0xFF
        assert sig.verify(TEST_MESSAGE, bytes(tampered_sig), pk) is False

    def test_empty_message(self):
        sig = Signature("ML-DSA-65")
        pk, sk = sig.keygen()
        signature = sig.sign(b"", sk)
        assert sig.verify(b"", signature, pk) is True

    def test_different_keypairs_produce_different_keys(self):
        sig = Signature("ML-DSA-65")
        pk1, sk1 = sig.keygen()
        pk2, sk2 = sig.keygen()
        assert pk1 != pk2
        assert sk1 != sk2


class TestSignatureAlgorithmEnumeration:
    """Test algorithm listing and query functions."""

    def test_sig_algorithms_not_empty(self):
        algs = sig_algorithms()
        assert len(algs) > 0

    def test_mldsa65_is_listed(self):
        algs = sig_algorithms()
        assert "ML-DSA-65" in algs

    def test_sig_is_enabled(self):
        assert sig_is_enabled("ML-DSA-65") is True
        assert sig_is_enabled("NONEXISTENT") is False
